jenkins/worker.py

```diff
@@ -0,0 +1,242 @@
+"""This is the script executed by workers of the quality control pipline."""
+
+import argparse
+import datetime
+import json
+from os.path import basename
+from pprint import pprint
+import re
+import subprocess
+
+from google.cloud import bigquery
+from google.cloud import datastore
+
+
+class Pattern(object):
+  """Defines a pattern for regular expression matching."""
+
+  def __init__(self, pattern):
+    self.regex = re.compile(pattern, re.MULTILINE)
+
+  # Returns a dictionary of named capture groups to extracted output. Returns an
+  # empty dict of no match was found.
+  def extract(self, inp):
+    match = self.regex.search(inp)
+    if match is None:
+      return {}
+    return match.groupdict()
+
+
+# Pattern matchers for the various fields of the '/usr/bin/time -v' output
+USER_TIME_PATTERN = Pattern(
+    r'^\s*User time \(seconds\): (?P<user_time>\d+.\d+|\d+)')
+SYSTEM_TIME_PATTERN = Pattern(
+    r'^\s*System time \(seconds\): (?P<system_time>\d+.\d+|\d+)')
+WALL_TIME_PATTERN = Pattern(
+    r'^\s*Elapsed \(wall clock\) time \(h:mm:ss or m:ss\): '
+    r'((?P<hours>\d{1,2}):|)(?P<minutes>\d{1,2}):(?P<seconds>\d{2}\.\d{2})')
+MAX_RES_SET_SIZE_PATTERN = Pattern(
+    r'^\s*Maximum resident set size \(kbytes\): (?P<max_set_size>\d+)')
+
+# Pattern matcher for extracting the HEAD commit SHA-1 hash.
+GIT_SHA1_PATTERN = Pattern(r'^(?P<sha1>[0-9a-f]{40})\s+HEAD')
+
+
+def get_head_git_sha1():
+  """Returns the SHA-1 hash of the commit tagged HEAD."""
+  output = subprocess.check_output([
+      'git', 'ls-remote',
+      'https://github.com/googlecartographer/cartographer.git'
+  ])
+  parsed = GIT_SHA1_PATTERN.extract(output)
+  return parsed['sha1']
+
+
+def extract_stats(inp):
+  """Returns a dictionary of stats."""
+  result = {}
+
+  parsed = USER_TIME_PATTERN.extract(inp)
+  result['user_time_secs'] = float(parsed['user_time'])
+
+  parsed = SYSTEM_TIME_PATTERN.extract(inp)
+  result['system_time_secs'] = float(parsed['system_time'])
+
+  parsed = WALL_TIME_PATTERN.extract(inp)
+  result['wall_time_secs'] = float(parsed.get('hours', 0.)) * 3600 + float(
+      parsed['minutes']) * 60 + float(parsed['seconds'])
+
+  parsed = MAX_RES_SET_SIZE_PATTERN.extract(inp)
+  result['max_set_size_kbytes'] = int(parsed['max_set_size'])
+
+  return result
+
+
+def retrieve_entity(datastore_client, kind, identifier):
+  """Convenience function for Datastore entity retrieval."""
+  key = datastore_client.key(kind, identifier)
+  return datastore_client.get(key)
+
+
+def create_job_selector(worker_id, num_workers):
+  """Constructs a round-robin job selector."""
+  return lambda job_id: job_id % num_workers == worker_id
+
+
+def run_cmd(cmd):
+  """Runs command both printing its stdout output and returning it as string."""
+  p = subprocess.Popen(
+      cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+  output = ''
+  while p.poll() is None:
```

Comment:
for line in p.stdout: should do the same thing or not? Poll should not be necessary. 

Also, collect lines into a list and '\n'.join() at the end.

Reply:
as discussed changed.

Reply:
Done